File: Python/day_5.py
def solve():
    with open("../inputs/day_5.txt", "r") as f:
        lines = f.readlines()

    data_lines = [[] for i in range(10)]
    data_lines_2 = [[] for i in range(10)]
    instructions = []

    for i, line in enumerate(lines):
        # first lines
        if i < 8:
            for j, char in enumerate(line):
                if (j - 1) % 4 == 0 and char != ' ':
                    data_lines[(j - 1) // 4 + 1].append(char)
                    data_lines_2[(j - 1) // 4 + 1].append(char)
        elif i > 9:
            temp = line[:-1].split(" ")
            instructions.append(temp)

    for line in data_lines:
        line.reverse()
    for line in data_lines_2:
        line.reverse()

    # part A
    for move in instructions:
        source, destination = int(move[3]), int(move[5])
        n = int(move[1])

        for i in range(n):
            char = data_lines[source].pop()
            data_lines[destination].append(char)

    # for line in data_lines:
    #     if line:
    #         print(line[-1])

    # part B
    for move in instructions:
        source, destination = int(move[3]), int(move[5])
        n = int(move[1])
        temp = []
        for i in range(n):
            char = data_lines_2[source].pop()
            temp.append(char)
        for i in range(n):
            char = temp.pop()
            data_lines_2[destination].append(char)

    for line in data_lines_2:
        if line:
            print(line[-1])


if __name__ == "__main__":
    solve()

File: Rust/src/day_5.rs


pub fn solution() {
    let mut input = include_str!("../../inputs/day_5.txt")
        .split_once("\n\n")
        .map(|(a, b)|
            (a.split("\n").collect::<Vec<&str>>(),
             b.split("\n")
                 .map(|n|
                     n.split(" ").skip(1).step_by(2)
                         .map(|m| m.parse::<i32>().unwrap())
                         .collect::<Vec<i32>>())
                 .collect::<Vec<Vec<i32>>>()))
        .unwrap(); input.1.pop();

    let mut stacc: Vec<Vec<char>> = vec![];
    for _ in 0..10 {
        stacc.push(vec![]);
    }

    // add values to stacc :]
    for row in input.0.iter().rev().skip(1) {
        for (j, char) in row.chars().enumerate() {
            if ((j as i32 - 1).rem_euclid(4) == 0) & (char != ' ') {
                stacc[j / 4 + 1].push(char);
            }
        }
    }

    let mut stacc_2_electric_boogaloo = stacc.to_vec();

    // do da poppin n pushing n stuff : )
    let mut temp = vec![];
    for instruction in input.1.iter() {
        for i in 0..instruction[0] {
            let moved_box = stacc[instruction[1] as usize].pop().unwrap();
            let moved_box_2 = stacc_2_electric_boogaloo[instruction[1] as usize].pop().unwrap();
            temp.push(moved_box_2);
            stacc[instruction[2] as usize].push(moved_box);
        }
        for _ in 0..temp.len() {
            let moved_box = temp.pop().unwrap();
            stacc_2_electric_boogaloo[instruction[2] as usize].push(moved_box);
        }
    }
    print!("Part 1 solution: ");
    for pile in stacc.iter().skip(1) {
        if let Some(char) = pile.last() {
            print!("{}", char);
        }
    } println!();

    print!("Part 2 solution: ");
    for pile in stacc_2_electric_boogaloo.iter().skip(1) {
        if let Some(char) = pile.last() {
            print!("{}", char);
        }
    }
}